### ISSUE TITLE
Raspberry Pi WS2812b implementation

### ISSUE THREAD

**Issue by chunk100** [SolidStateGroup/rio#1]
This isnt an issue, more of a plan for implementation as requested in your readme.  You were featured in the [Raspberry Pi weekly](https://www.raspberrypi.org/weekly/five/) which is where if found this.

I am at the stage you were to begin with and have completed a 10x15 grid using the Raspberry pi neopixels library but also a Fadecandy board (seemed to be more reliable).  This was always just the beginning and the next plan was to go large and somewhat topically BUILD A WALL!

Before I do, I hope to implement your library on my 10x15 grid.  This will no doubt take some thinking and reading of what you have here.  In the meantime, do you have a bit more of an idiots guide to implementing this on the Raspberry Pi?

You may find the [Glediator software](http://www.solderlab.de/index.php/software/glediator) interesting and fun!

**Comment by kyle-ssg** [comment]
Hey @chunk100, thanks for letting us know about Raspberry Pi weekly - that's pretty cool! Apologies for the late reply, a lot of us were out at SXSW.

Would be glad to help you out here, I'm pretty keen to see what people can build with this. Are you looking for more of a guide on the hardware side or software? Anything that isn't obvious I'll write up and add to the Read me.

If you've run rio node server you will notice that the console starts showing a grid of leds, there's configuration <a href="https://github.com/SolidStateGroup/rio/blob/master/rio-server/config/index.js">here</a> to output everything you see on the console to a pi that is hooked up to.

If you want to output to the pi you set ```sendToPi:true``` and connect your LEDS up to the GPIO pins 18 and 19 and run ```python ./firmware/rpi/server.py``` alongside your node server. The firmware and PIN configuration can be found <a href="https://github.com/SolidStateGroup/rio/blob/master/rio-server/firmware/rpi/server.py">Here</a>. Assuming you can run a <a href="https://github.com/SolidStateGroup/rio/blob/master/rio-server/firmware/rpi/strandtest.py">strand test</a> it should all work fine.

**Comment by chunk100** [comment]
Thanks for the reply, it all looks great.

Starting at the end, I have got the strandtest.py running but using the
great adafruit guide

https://learn.adafruit.com/neopixels-on-raspberry-pi/overview

There have been reports of issues with the Raspberry Pi 3 which I have also
encountered but no problems with the older Pis.  This comes with a
disadvantage of not having the built in wifi etc which the later models
do.  My solution was to use the Fadecandy board with the Raspberry Pi:

https://learn.adafruit.com/1500-neopixel-led-curtain-with-raspberry-pi-fadecandy/overview

There is a python implementation of Fadecandy which I am using but also
Java, Processing etc.  I'm pretty OK with the hardware and understand you
are using the PWM implementation from pin 18.

My experience for the rest of it is limited!  I have so far managed to
clone the library and that's it!  Presumable I run the server on the
Raspberry Pi and then somehow tell it what kind of grid I have.  My Java
experience is limited but a guide along the lines of the adafruit one would
be great and there would no doubt be a big uptake!  I'd hopefully get this
working with my 10x15 grid before then venturing into something substantial

**Comment by kyle-ssg** [comment]
No problem for the reply. Luckily the software side is something I can definitely help with  😄 . Getting the strand test working is a good start, that way we can concentrate on getting the python firmware and rio-server all configured to your matrix. You won't really have to do that much to get this up and running.

Assuming you have ran the commands in the installation section of the readme, here's what you need to do:

**Step 1 - configure your matrix dimensions**
Change ```matrix``` in <a href="https://github.com/SolidStateGroup/rio/blob/master/rio-server/config/index.js">rio-server/config/index.js</a> to your width and height. 

**Step 2 - verify that the output is correct**
Running rio-server (``npm run server``) will output a gif to the console at 10x15px. You can change the gif by editing <a href="https://github.com/SolidStateGroup/rio/blob/master/rio-server/index.js#L44">rio-server/index.js</a>.  

**Step 3 - configure firmware and set server to send to pi**
Set ```sendToPi: true``` in <a href="https://github.com/SolidStateGroup/rio/blob/master/rio-server/config/index.js#L5">rio-server/config/index.js</a>. (This will basically write all pixel data seen in the console to a stream which is picked up by the python firmware). 

Change the firmware to output to the correct number of pixels
<a href="https://github.com/SolidStateGroup/rio/blob/master/rio-server/firmware/rpi/server.py#L60">rio-server/firmware/rpi/server.py</a>

- set ```conn.recv(6120)``` to ```conn.recv(450)```  (the number of LEDs x3 - r,g,b)
- set ```LED_COUNT      = 1020 ``` to ```LED_COUNT      = 150 ```  (the number of LEDs)

**Step 4 - run node server and python firmware on the pi**
```npm run server```
 ```python ./firmware/rpi/server.py```

**Comment by chunk100** [comment]
So I started through the readme but stumbled pretty early on.  I spent the day trying to install Linuxbrew on my Raspberry Pi which seemed to work, using

 

http://linuxbrew.sh/

 

However, brew doesn’t seem to be working as when I tried to test it, I got this:

 

pi@mypi:~ $ brew install hello
==> Downloading https://ftpmirror.gnu.org/hello/hello-2.10.tar.gz
==> Downloading from http://reflection.oss.ou.edu/gnu/hello/ ... .10.tar.gz <http://reflection.oss.ou.edu/gnu/hello/hello-2.10.tar.gz> 
######################################################################## 100.0%
==> ./configure --disable-silent-rules --prefix=/home/pi/.linuxbrew/Cellar/hello/2.10
*** Error in `/usr/bin/gcc-4.9': double free or corruption (top): 0x006d5718 ***
Last 15 lines from /home/pi/.cache/Homebrew/Logs/hello/01.configure:
--prefix=/home/pi/.linuxbrew/Cellar/hello/2.10

configure: WARNING: unrecognized options: --disable-debug
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... /usr/bin/gcc-4.9
checking whether the C compiler works... no
configure: error: in `/tmp/hello-20170318-13259-yczq0w/hello-2.10':
configure: error: C compiler cannot create executables
See `config.log' for more details

READ THIS: https://github.com/Linuxbrew/brew/blob/ ... leshooting <https://github.com/Linuxbrew/brew/blob/master/docs/Troubleshooting.md#troubleshooting> 

 

Various forums have revealed that:

 

“Currently only your host GCC compiler and gcc-5 are supported on Linuxbrew. We hope to add support for the GCC 4.x and GCC 6 in the future.”

 

Can you confirm you got Linuxbrew running OK on a Raspberr Pi?  If so, any ideas how?

 

If brew is a no go, any alternatives to installing:

 

brew install pkg-config cairo libpng jpeg giflib

**Comment by lukefanning** [comment]
Hi @chunk100, I am another engineer who worked on rio. Yep you shouldn't need Linuxbrew to install those packages. Instead run `apt-get install pkg-config cairo libpng jpeg giflib` and it should pick them up. You might need to upgrade your pi from Wheezy to Jessie if you start running into GCC compiler errors when installing those packages. I'm fairly certain I was forced to do so for either those packages or ffmpeg codecs support but can't remember which. Perhaps let me know the error before embarking on any sort of OS upgrade.

**Comment by chunk100** [comment]
I think I am slowly getting there.  The apt-get install only allows for pkg-config, the other 4 have to be hunted for but I think I have got them installed.

 

I then went to run the server but had to install npm which was successful.

 

When then running the npm server, I get:

 

pi@mipi:~/rio$ cd ./rio-server && npm i

npm WARN package.json rio-server@1.0.0 No repository field.

npm WARN deprecated node-uuid@1.4.7: use uuid module instead

npm ERR! 404 Not Found

npm ERR! 404

npm ERR! 404 'slack/client' is not in the npm registry.

npm ERR! 404 You should bug the author to publish it

npm ERR! 404 It was specified as a dependency of 'rio-server'

npm ERR! 404

npm ERR! 404 Note that you can also install from a

npm ERR! 404 tarball, folder, or http url, or git url.

 

npm ERR! System Linux 4.4.50-v7+

npm ERR! command "/usr/bin/nodejs" "/usr/bin/npm" "i"

npm ERR! cwd /home/pi/rio/rio-server

npm ERR! node -v v0.10.29

npm ERR! npm -v 1.4.21

npm ERR! code E404

npm ERR!

npm ERR! Additional logging details can be found in:

npm ERR!     /home/pi/rio/rio-server/npm-debug.log

npm ERR! not ok code 0

 

Any ideas??

**Comment by lukefanning** [comment]
Ok great nearly there! You aren't actually running the server with that command `npm i` but you are attempting to install rio-server's dependencies with it. You're running a very old version of Node (+ npm). You'll need to upgrade to at least version 6. I now recall this can be a cause for needing an OS upgrade to Jessie. Try the following:

```
curl -sL https://deb.nodesource.com/setup_6.x | sudo -E bash -
sudo apt-get install nodejs
```

then run `node -v` and see whether it gives you back a version number starting with 6. If you get a nasty looking GCC error it looks like you'll be needing to upgrade the OS unfortunately.

**Comment by chunk100** [comment]
I think I am getting there.. slowly!  I am running the latest (March 2017)
Raspbian Jessie distribution.
On the Raspberry Pi, typing:

apt-get install pkg-config cairo libpng jpeg giflib

only installs pkg-config, the rest cannot be found.  I managed to do
them all manually which was a bit of a work around which Ill document
for you sometime if you want.

I decided to start afresh and am about to install Node using your
instructions.  I see the latest version is 7 and you are showing 6.
Does it matter about the version number?

**Comment by lukefanning** [comment]
As long as its at least version 6 then it should be fine. I recommend v6 simply because it is very stable and contains all the ES6 features etc required for rio-server to run.

Regarding the node-canvas dependencies, I don't remember it being that complicated.. I think its just that the package names are different on the apt repo. Would appreciate any documentation you're willing to give once you have it running. I didn't write enough down!

**Comment by chunk100** [comment]
For installation of the prerequisites , I did:


Giflib


https://sourceforge.net/p/giflib/discussion/352789/thread/58ae0159/

download latest giflib

extract and copy folder to raspberry pi

Instal autoconf and libtool
$ sudo apt-get install autoconf
$ sudo apt-get install libtool

to compile static lib of giflib-5.1.4:
$ cd /home/pi/giflib-5.1.4
$  bash autogen.sh --prefix=/home/pi/giflib-5.1.4 --enable-static --disable-shared
$ make
$ make check
$ make install


Libpng


https://geeksww.com/tutorials/libraries/libpng/installation/installing_libpng_on_ubuntu_linux.php


Cairo


sudo apt-get install libcairo2-dev


jpeg


http://stackoverflow.com/questions/11969000/how-to-install-libjpeg

did all this:

sudo apt-get install libjpeg-dev

sudo apt-get install zlib1g-dev

sudo apt-get install libpng-dev

I then installed Node 6 as recommended and this all went OK.

I then got to:

cd ./rio-server && npm i

This seemed to go well although towards the end, there were a number of errors.  I hope you can make sense of them:

pi@mypi:~/rio $ cd ./rio-server && npm i

npm WARN deprecated node-uuid@1.4.7: use uuid module instead

 

> canvas@1.6.5 install /home/pi/rio/rio-server/node_modules/canvas

> node-gyp rebuild

 

make: Entering directory '/home/pi/rio/rio-server/node_modules/canvas/build'

  SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node

  COPY Release/canvas-postbuild.node

  CXX(target) Release/obj.target/canvas/src/Canvas.o

  CXX(target) Release/obj.target/canvas/src/CanvasGradient.o

  CXX(target) Release/obj.target/canvas/src/CanvasPattern.o

In file included from ../src/CanvasPattern.cc:9:0:

../src/Image.h:19:21: fatal error: gif_lib.h: No such file or directory

#include <gif_lib.h>

                     ^

compilation terminated.

canvas.target.mk:118: recipe for target 'Release/obj.target/canvas/src/CanvasPattern.o' failed

make: *** [Release/obj.target/canvas/src/CanvasPattern.o] Error 1

make: Leaving directory '/home/pi/rio/rio-server/node_modules/canvas/build'

gyp ERR! build error

gyp ERR! stack Error: `make` failed with exit code: 2

gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)

gyp ERR! stack     at emitTwo (events.js:106:13)

gyp ERR! stack     at ChildProcess.emit (events.js:191:7)

gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)

gyp ERR! System Linux 4.4.50-v7+

gyp ERR! command "/usr/bin/nodejs" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"

gyp ERR! cwd /home/pi/rio/rio-server/node_modules/canvas

gyp ERR! node -v v6.10.0

gyp ERR! node-gyp -v v3.4.0

gyp ERR! not ok

rio-server@1.0.0 /home/pi/rio/rio-server

├─┬ @slack/client@3.9.0

│ ├── async@1.5.2

│ ├── bluebird@3.5.0

│ ├── eventemitter3@1.2.0

│ ├─┬ https-proxy-agent@1.0.0

│ │ ├─┬ agent-base@2.0.1

│ │ │ └── semver@5.0.3

│ │ └── debug@2.6.3

│ ├── inherits@2.0.3

│ ├── pkginfo@0.4.0

│ ├── retry@0.9.0

│ ├── url-join@0.0.1

│ └─┬ winston@2.3.1

│   ├── async@1.0.0

│   ├── cycle@1.0.3

│   ├── eyes@0.1.8

│   └── stack-trace@0.0.9

├── ansi-256-colors@1.1.0

├─┬ body-parser@1.17.1

│ ├── bytes@2.4.0

│ ├── content-type@1.0.2

│ ├─┬ debug@2.6.1

│ │ └── ms@0.7.2

│ ├── depd@1.1.0

│ ├── http-errors@1.6.1

│ ├── iconv-lite@0.4.15

│ ├─┬ on-finished@2.3.0

│ │ └── ee-first@1.1.1

│ ├── qs@6.4.0

│ ├─┬ raw-body@2.2.0

│ │ └── unpipe@1.0.0

│ └─┬ type-is@1.6.14

│   └── media-typer@0.3.0

├─┬ cors@2.8.1

│ └── vary@1.1.0

├─┬ express@4.15.2

│ ├─┬ accepts@1.3.3

│ │ └── negotiator@0.6.1

│ ├── array-flatten@1.1.1

│ ├── content-disposition@0.5.2

│ ├── cookie@0.3.1

│ ├── cookie-signature@1.0.6

│ ├── debug@2.6.1

│ ├── encodeurl@1.0.1

│ ├── escape-html@1.0.3

│ ├── etag@1.8.0

│ ├─┬ finalhandler@1.0.0

│ │ └── debug@2.6.1

│ ├── fresh@0.5.0

│ ├── merge-descriptors@1.0.1

│ ├── methods@1.1.2

│ ├── parseurl@1.3.1

│ ├── path-to-regexp@0.1.7

│ ├─┬ proxy-addr@1.1.3

│ │ ├── forwarded@0.1.0

│ │ └── ipaddr.js@1.2.0

│ ├── range-parser@1.2.0

│ ├─┬ send@0.15.1

│ │ ├── debug@2.6.1

│ │ └── destroy@1.0.4

│ ├── serve-static@1.12.1

│ ├── setprototypeof@1.0.3

│ ├── statuses@1.3.1

│ └── utils-merge@1.0.0

├─┬ fluent-ffmpeg@2.1.0

│ └─┬ which@1.2.12

│   └── isexe@1.1.2

├── lodash@4.17.4

├─┬ node-fetch@1.6.3

│ ├── encoding@0.1.12

│ └── is-stream@1.1.0

├─┬ node-ipc@8.10.3

│ ├── colors@1.0.3

│ ├── event-pubsub@4.2.3

│ ├── js-message@1.0.5

│ └─┬ js-queue@2.0.0

│   └── easy-stack@1.0.0

├── node-uuid@1.4.7

├─┬ request@2.81.0

│ ├── aws-sign2@0.6.0

│ ├── aws4@1.6.0

│ ├── caseless@0.12.0

│ ├─┬ combined-stream@1.0.5

│ │ └── delayed-stream@1.0.0

│ ├── extend@3.0.0

│ ├── forever-agent@0.6.1

│ ├─┬ form-data@2.1.2

│ │ └── asynckit@0.4.0

│ ├─┬ har-validator@4.2.1

│ │ ├─┬ ajv@4.11.5

│ │ │ ├── co@4.6.0

│ │ │ └─┬ json-stable-stringify@1.0.1

│ │ │   └── jsonify@0.0.0

│ │ └── har-schema@1.0.5

│ ├─┬ hawk@3.1.3

│ │ ├── boom@2.10.1

│ │ ├── cryptiles@2.0.5

│ │ ├── hoek@2.16.3

│ │ └── sntp@1.0.9

│ ├─┬ http-signature@1.1.1

│ │ ├── assert-plus@0.2.0

│ │ ├─┬ jsprim@1.4.0

│ │ │ ├── assert-plus@1.0.0

│ │ │ ├── extsprintf@1.0.2

│ │ │ ├── json-schema@0.2.3

│ │ │ └── verror@1.3.6

│ │ └─┬ sshpk@1.11.0

│ │   ├── asn1@0.2.3

│ │   ├── assert-plus@1.0.0

│ │   ├── bcrypt-pbkdf@1.0.1

│ │   ├─┬ dashdash@1.14.1

│ │   │ └── assert-plus@1.0.0

│ │   ├── ecc-jsbn@0.1.1

│ │   ├─┬ getpass@0.1.6

│ │   │ └── assert-plus@1.0.0

│ │   ├── jodid25519@1.0.2

│ │   ├── jsbn@0.1.1

│ │   └── tweetnacl@0.14.5

│ ├── is-typedarray@1.0.0

│ ├── isstream@0.1.2

│ ├── json-stringify-safe@5.0.1

│ ├─┬ mime-types@2.1.14

│ │ └── mime-db@1.26.0

│ ├── oauth-sign@0.8.2

│ ├── performance-now@0.2.0

│ ├── safe-buffer@5.0.1

│ ├── stringstream@0.0.5

│ ├─┬ tough-cookie@2.3.2

│ │ └── punycode@1.4.1

│ ├── tunnel-agent@0.6.0

│ └── uuid@3.0.1

├─┬ twit@2.2.5

│ └── mime@1.3.4

├─┬ ws@1.1.4

│ ├── options@0.0.6

│ └── ultron@1.0.2

└─┬ youtube-dl@1.11.1

  ├─┬ mkdirp@0.5.1

  │ └── minimist@0.0.8

  └─┬ streamify@0.2.8

    └─┬ hashish@0.0.4

      └── traverse@0.6.6

 

npm WARN rio-server@1.0.0 No repository field.

npm WARN rio-server@1.0.0 No license field.

npm ERR! Linux 4.4.50-v7+

npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "i"

npm ERR! node v6.10.0

npm ERR! npm  v3.10.10

npm ERR! code ELIFECYCLE

 

npm ERR! canvas@1.6.5 install: `node-gyp rebuild`

npm ERR! Exit status 1

npm ERR!

npm ERR! Failed at the canvas@1.6.5 install script 'node-gyp rebuild'.

npm ERR! Make sure you have the latest version of node.js and npm installed.

npm ERR! If you do, this is most likely a problem with the canvas package,

npm ERR! not with npm itself.

npm ERR! Tell the author that this fails on your system:

npm ERR!     node-gyp rebuild

npm ERR! You can get information on how to open an issue for this project with:

npm ERR!     npm bugs canvas

npm ERR! Or if that isn't available, you can get their info via:

npm ERR!     npm owner ls canvas

npm ERR! There is likely additional logging output above.

 

npm ERR! Please include the following file with any support request:

npm ERR!     /home/pi/rio/rio-server/npm-debug.log

**Comment by chunk100** [comment]
As it turns out, I also need 

Sudo apt-get giflib-dbg

This then gives this which looks better!

pi@mypi:~/rio $ cd ./rio-server && npm i

> canvas@1.6.5 install /home/pi/rio/rio-server/node_modules/canvas

> node-gyp rebuild

make: Entering directory '/home/pi/rio/rio-server/node_modules/canvas/build'

  SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node

  COPY Release/canvas-postbuild.node

  CXX(target) Release/obj.target/canvas/src/Canvas.o

  CXX(target) Release/obj.target/canvas/src/CanvasGradient.o

  CXX(target) Release/obj.target/canvas/src/CanvasPattern.o

  CXX(target) Release/obj.target/canvas/src/CanvasRenderingContext2d.o

  CXX(target) Release/obj.target/canvas/src/color.o

  CXX(target) Release/obj.target/canvas/src/Image.o

../src/Image.cc:664:13: warning: ‘void jpeg_mem_src(j_decompress_ptr, void*, long int)’ defined but not used [-Wunused-function]

static void jpeg_mem_src (j_decompress_ptr cinfo, void* buffer, long nbytes) {

             ^

  CXX(target) Release/obj.target/canvas/src/ImageData.o

  CXX(target) Release/obj.target/canvas/src/init.o

  CXX(target) Release/obj.target/canvas/src/FontFace.o

  SOLINK_MODULE(target) Release/obj.target/canvas.node

  COPY Release/canvas.node

make: Leaving directory '/home/pi/rio/rio-server/node_modules/canvas/build'

rio-server@1.0.0 /home/pi/rio/rio-server

└─┬ canvas@1.6.5

  ├── nan@2.5.1

  ├─┬ parse-css-font@2.0.2

  │ ├── css-font-size-keywords@1.0.0

  │ ├── css-font-stretch-keywords@1.0.1

  │ ├── css-font-style-keywords@1.0.1

  │ ├── css-font-weight-keywords@1.0.0

  │ ├── css-global-keywords@1.0.1

  │ ├── css-list-helpers@1.0.1

  │ ├── css-system-font-keywords@1.0.0

  │ ├── tcomb@2.7.0

  │ └── unquote@1.1.0

  └─┬ units-css@0.4.0

    ├── isnumeric@0.2.0

    └── viewport-dimensions@0.2.0

npm WARN rio-server@1.0.0 No repository field.

npm WARN rio-server@1.0.0 No license field.

**Comment by chunk100** [comment]
Turns out I also needed 
sudo apt-get giflib-dbg

Ill try running it later today ;)

**Comment by lukefanning** [comment]
Yep another dependency needed for node-canvas to install correctly. You need to run `sudo apt-get install libgif-dev` and then re-run `npm i`. As for the error you're getting on running the server with `npm run start` I am actually seeing the same thing. I think there's been a change in the latest version of node-canvas. Will get back to you on a solution

**Comment by lukefanning** [comment]
If you do a pull on rio-server, I have frozen the `canvas` version to `1.6.3` which is the last version that actually works. Try again with that. Also the node-canvas README did have a list of the Linux based dependencies `sudo apt-get install libcairo2-dev libjpeg8-dev libpango1.0-dev libgif-dev build-essential g++`.

**Comment by chunk100** [comment]
Tried again, it took a long time but seemed to work without any errors.  I took lots of notes so hope to write it all up sometime.  I then edited the matrix size as above.

I then ran the server (using sudo) and got a few error messages but the screen then came to life as a flickering fragment of the gif.  However.... (see below!)

pi@mypi:~/rio $sudo npm run server
> rio@0.0.1 server /home/pi/rio
> cd ./rio-server && npm start


> rio-server@1.0.0 start /home/pi/rio/rio-server
> node index.js

[SECURITY] node-uuid: crypto not usable, falling back to insecure Math.random()
error: not_authed
error: unrecoverable failure connecting to the RTM API: not_authed

I edited the server.py file to my 10x15 grid and also got rid of strip2 as not required.  For infor, suspect also need to edit the line:

# Check whether end of frame has been reached and if so reset pixel index and show pixels
                    if index == 10 * 15:#

Then run the server.py in a different Terminal window
sudo python server.py
Opening socket...
Listening...
accepted connection

And thats where it stops.  I tested the grid using a simple strandtest program and it works, however, dont appear to be lighting any lights :(

**Comment by lukefanning** [comment]
Hi @chunk100 Have you enabled sendToPi in `config/index.js` in rio-server? And while you're in there update the matrix size as well if you haven't done so already. Make sure to start the python script before running the node side. At some point we plan to change rio-server so that it starts the python script from within the node process.

As for that error, its probably coming from the initialisation of the Twitter integration. It would normally need to be configured in `config/index.js`. It should be optional and we'll look into that. I don't think it prevents the server from sending image data though.

**Comment by chunk100** [comment]
I had another look at this today but still cant make progress.  The index.js file has sendToPi set to true and all else works but still get stuck on the 'accepted connection'.

Looking at the server.py file,  it seems to be getting stuck on the line data = conn.recv(450) line.  I used 450 as assumed this is3x the total number of pixels to cover R G and B.

Is there any way to see what the server.py file is receiving if anything?

**Comment by ssg-luke** [comment]
Yep its actually a requirement to run both on the same RPi as they use a IPC socket to communicate. You could add in a `print "Calling show on strips"` before `strip.show()` in `server.py` to make sure that the script is receiving pixel data from rio-server. Its expected that `Accepted connection` is the last thing you'll see before it starts lighting up LEDs. Try running `singlestrandtest.py` after editing the values at the top for your LED grid. If that works as expected then send me your `server.py` so I can double check it. I know that you would have had to have edited it for your single strip.

**Comment by chunk100** [comment]
Thanks for the reply.  I did try what you suggested and even added some other troubleshooting including a quick test of the pixels prior to any of the 'Open socket' lines.  I also added a line print("received") just after the 'data = conn.recv(450)' which is never displayed so it is getting stuck on this 'data = conn.recv(450)' line.

Im happy to send the server.py file if it will help, just let me know where

**Comment by ssg-luke** [comment]
Yep to tell if data is being sent in the first place you will need to add some console logging to `outputs/pi-output.js`. In drawFrame() add `console.log(channel ? 'Connected to RPi and drawing frame' : 'Not connected to RPi. Cannot draw frame')`. Yes I would expect that both rio-server and the python script would both be run as root. If you aren't running the former as root that might be why it's not working as both sides require access to the /tmp folder on your RPi. Will make sure to add that to the rio instructions if it solves your issue. I'll have a think about why the python script seems to move on after stopping rio. Sounds like rio-server is holding up the pipe hopefully because its not being run as root.

**Comment by chunk100** [comment]
I have been running it all as root, one of the first things tried actually - it makes a difference for the server.py but not the rio server.  Its worth adding to the instructions anyway.

I edited pi-output.js and drawFrame() now looks like this:

    drawFrame: function(frame, cb) {
        callback = cb;
        channel && channel.emit(frame);
        console.log(channel ? 'Connected to RPi and drawing frame' : 'Not connected to RPi. Cannot draw frame');
    },

I ran everything again and nothing changed, have I done it right?

Just out of curiosity, have you tried installing this on a Pi from a fresh install?!!

**Comment by ssg-luke** [comment]
We haven't installed rio on a fresh RPi yet. Looks like it is worth doing :).

Ok so you don't see that console.log() message at all in rio-server's output? That either indicates nothing is actually being sent to the pi-output or its locking up on the IPC stuff which would be unexpected. Can you confirm that you've left the current startup GIF in `rio-server/index.js` as the default i.e. the rio GIF. Does your RPi have internet access? That would be required for download of the GIF. Can you run `ls /tmp` on your RPi and confirm whether you see a app.main file in there?

**Comment by chunk100** [comment]
Havnt touched the gif, its the same one.  Pi is also attached to internet

app.main is in the /tmp directory.

Still nothing coming through on server.py

**Comment by ssg-luke** [comment]
Can you put a `console.log('Drawing frame');` at the top of the `drawFrame()` func in `pi-output.js` and re run rio-server? Leave the other logging message in there. Need to be sure about whether rio-server ever calls this function or not. I think its failing on the `channel.emit()` line (but I don't know why) and Node.js doesn't necessarily always produce error output unless capturing all uncaught exceptions. I wouldn't mind taking a direct look at your RPi over SSH if you were willing to give me appropriate credentials (i.e. with superuser privileges). Just send me a PM if you're comfortable doing that.

**Comment by chunk100** [comment]
I added that line but it didnt change anything.  Im not sure what else to try, maybe a fresh install all round?!!

**Comment by Chilus** [comment]
I'm having similar issues as chun100, where running the command: `npm run server` won't execute as many dependencies are missing. Could you have a look at this by trying to run the library on a fresh Pi3 install?

**Comment by kyle-ssg** [comment]
@Chilus @chunk100 Sorry for the late reply, we've added some Pi specific installation instructions here
https://github.com/SolidStateGroup/rio/tree/master/rio-server. Doing this we were able to install on a fresh Pi and Raspbian VM within a couple minutes. 

This topic has deviated to a few different things, I'm going to close this in the hope that the added instructions are of help. If you still face the same issue as the original description then reopen this issue, otherwise, it'd probably be less convoluted to open a new one with your specific query / issue.
